[WIP] Add command parser

lupoDharkael opened the pull request:
Initial work for https://github.com/godotengine/godot/issues/20000

Proposal: https://github.com/godotengine/godot-proposals/issues/137

The parser is pretty simple. We have the class `CommandFlag` which represents a single commands and its properties (the code is self explanatory). The class `CommandParser` contains the list of `CommandFlag` and process the arguments in the `parse` method.
The method `init_defaults` adds all the commands of the engine.
With a centralised point with all the data of the commands  the --help can be generated dynamically. This has some advantages such as proper alignment of descriptions based on the largest line, and even splitting the lines if they exceed the 80 characters.

Engine arguments and project arguments are separated by a -- as it was suggested in the issue:
godot --godot-command -- project-command 333 4

TODO:
- [x] Implement a way to add `CommandFlag` checkers from gdscript.
- [x] Method name cleanup.
- [ ] Decide about List vs Vector internal storage.
- [ ] Add class documentation.

neikeq commented on main/cli_parser.cpp:
```diff
@@ -0,0 +1,633 @@
+	resolution_command.set_data("Request window resolution.", "Display options");
+	resolution_command.set_argument_name("<W>x<H>");
+	resolution_command.add_checker(check_resolution, "Invalid resolution, it should be e.g. '1280x720' and the values must be above 0.");
+	add_command(help_command);
```
All the previous `add_command` calls are passing `help_command`, I guess by mistake.

neikeq commented on main/cli_parser.cpp:
```diff
@@ -0,0 +1,633 @@
+	test_description = ").";
+	gdnative_generate_json_api_command.set_data(test_description, "Standalone options");
+	gdnative_generate_json_api_command.set_argument_name("<test>");
+	add_command(help_command);
```
Same for a lot of the previous commands.

neikeq commented:
Didn't read the description, I guess this is still WIP. This is great any ways and it would be great if the main `CommandParser` could be publicly exposed to make it possible to add custom commands from modules. This would specially benefit the mono module, which has some hard-coded commands.

lupoDharkael commented:
I'm adding a [WIP] to the title of the PR so it is more clear.
@neikeq I need to change all the copypasted parts like the ones you noticed. Godot has a ton of arguments and it was tedious creating every single one.
This PR is still very WIP. I just pushed it so it can receive some feedback.

Calinou commented:
Nice work! :slightly_smiling_face:

Would it be possible to expose this to GDScript so that it can be used direclty in projects? I can see it being useful to quickly jump to a specific level in a game (or similar).

lupoDharkael commented:
@Calinou I'll leave the potential GDScript API for a future issue so it can be debated. It would require some minor tweaks in the parser.

@neikeq having the modules register their commands would be ideal but I'm a little lost about how it could be done as it would require the registration to be ready early on `Main::setup`.
Maybe the register* functions like register_mono_types() could be used as the point where commands are defined.
The problem is, it would be good to notify when the parser has processed the cms args so the modules can initialise values based on the results.
Any hints would be really helpful.

neikeq commented:
Hmm, It's indeed not that simple since command line is parsed before calling `register_module_types` and I don't think moving the latter is a good idea, so it would require a custom callback for initializing command line options.
I wouldn't bother too much with it for now. It can be done in a new PR after thinking this well.

lupoDharkael commented:
Now Godot suggests commands if you wrote them wrong:

![image](https://user-images.githubusercontent.com/14951430/53512057-20cec900-3ac2-11e9-861c-19b99c4421b4.png)

lupoDharkael commented:
I've exposed the argument check functions so they can be reused when the command registration in modules is implemented.

The PR is practically ready for testing and review.

lupoDharkael commented on core/bind/core_bind.cpp:
```diff
@@ -483,7 +483,7 @@ String _OS::get_name() const {
 Vector<String> _OS::get_cmdline_args() {
 
-	List<String> cmdline = OS::get_singleton()->get_cmdline_args();
+	List<String> cmdline = OS::get_singleton()->get_project_cmdline_args();
```
I'm not sure if it should be like this. This way calling OS.get_cmdline_args() from gds would return the arguments for the game.
For example in: `godot --quiet -- test 55 3`
get_cmdline_args() would return` ["test", "55", "3"]`

But the docs say get_cmdline_args returns "the command line arguments passed to the engine".
Maybe it should just pass all the commands. The alternative is exposing `get_project_cmdline_args` (which I implemented in the class OS) so we have both, all the arguments game+engine (get_cmdline_args) and game arguments only (get_project_cmdline_args).

Xrayez replied:
This could break compat imo, there could be some editor tools which rely on engine-specific args, I'd just expose `get_project_cmdline_args` separately from `get_cmdline_args`. Yet I'm still not sure how these project-specific args get registered by the parser currently.

> Maybe it should just pass all the commands.

This could be a decent trade-off, + exposing `CommandParser` to scripting would help developers to sort out engine-specific from project-specific args. What about "black-listing" (or rather reserving) engine-specific args?

Xrayez replied:
As a side note, here's how I emulate project-specific args from within project settings, editor-only:

![godot_main_loop_args](https://user-images.githubusercontent.com/17108460/65877410-4b546d80-e394-11e9-912b-1ffa675d6ae7.png)

lupoDharkael commented:
Ok so this is ready for testing. The get_cmdline_args thin still needs feedback. Once this is merged I'll start working on the PR so godot modules can register their commands in the parser.

lupoDharkael commented:
@akien-mga I'm reworking this PR based on the feedback of the proposal and this issue and I'm adapting the code I didn't touch in the actual state of the PR (for example the usage of cmdline in the engine tests)

I have a doubt about a check condition you implemented:
https://github.com/godotengine/godot/blob/5ca74604217758d91e344b052fe87c9eda8940a9/main/tests/test_math.cpp#L489-L496

What should that `file` be? a gds script file? I don't think I'm understanding the usage of that test.

lupoDharkael commented:
I also think this PR should be moved to the 4.0 milestone.

akien-mga commented:
Yeah that's 4.0 material indeed, I just didn't want to change milestone before taking time to review any of the proposed changes... But time is scarce :)

I fully support the idea to refactor our command line interface as it's currently very limited and unforgiving. Haven't checked this PR's code yet so I can't comment on the implementation, but it's likely going in the right direction :)

> What should that test file be? a gds script file? I don't think I'm understanding the usage of that test.

I didn't implement this test, I only fixed up some issues. I also have no clue about what kind of files is expected here, but I guess it used to be a custom GDScript file with some math stuff to evaluate.

These tests are also obsolete and will be replaced in the future, so I wouldn't bother too much about the `--test` interface in this PR. See #30743 for the likely replacement.

lupoDharkael commented:
I think I'll need to change the cmdline from List<String> to Vector<String> so I can prevent creating an extra class to expose to gds. I don't think keeping it as List<String> makes sense anyways as the cmdline doesn't change and exposing them to gds requires copying them to a Vector<String> anyways.

Xrayez commented on core/os/os.h:
```diff
@@ -282,8 +284,8 @@ class OS {
-	virtual List<String> get_cmdline_args() const { return _cmdline; }
-	virtual List<String> get_game_args() const { return _project_cmdline; }
+	virtual List<String> get_cmdline_args() const { return _command_parser->get_cmdline_args(); }
+	virtual List<String> get_game_args() const { return _command_parser->get_project_args(); }
```
Could name it as `get_project_args` as well, see godotengine/godot-proposals#190. 🙂

lupoDharkael replied:
Done! thanks for the suggestion!

lupoDharkael commented:
I've exposed the parser to gdscript! I'm doing some cleanup so I'll push the commit later.
Here's a screenshot of it working:
![image](https://user-images.githubusercontent.com/14951430/68543837-7fbc4000-03bc-11ea-87bb-83059281eaf7.png)

lupoDharkael commented on core/cli_parser.cpp:
```diff
@@ -0,0 +1,943 @@
+
+	StringName func;
+	Object *obj;
+};
```
The class used to check the argument passed after a command is polymorphic because it can be a pointer to function (for easier in-engine development), a pointer to Object + StringName (to define checkers from GDScript). A lambda checker could be added in the future as it is a possibility that references to script functions will be added to Variant.

lupoDharkael committed:
Add command parser

lupoDharkael committed:
Add command parser

rainlizard commented:
Hi, will this solve #23744 ?

Godot currently cannot be used to create software applications that associate files. For example, if you want to create a simple image viewer, you will be unable to associate PNG files with the executable. Because double clicking a PNG file will open the program and then crash with "No loader found for resource" error.

lupoDharkael commented:
This needs a rebase (again). I'll update this in a few days now that vulkan has been merged!

aaronfranke commented:
@lupoDharkael The majority of the core refactoring work done by the core devs is finished, so anytime soon would be a good time to rebase this.

lupoDharkael commented:
I was waiting because I expected a lot of things to be changed not knowing when will this be considered to be reviewed more in depth.
I'm updating this soon™️ (it may take a while because I need to improve a few things, not only fix the conflicts).

Anutrix commented:
Requesting rebase.

lupoDharkael commented:
Oh I forgot about this one. I'll try to find time to resume this PR

Shatur95 commented:
@lupoDharkael, are you still interested in it? Maybe some of us can just pick this PR and continue working on it?

lupoDharkael commented:
I don't know if I'll be able to resume this work soon as I have university exams and other stuff going on. I wouldn't mind If someone would like to continue my work.

Shatur95 commented:
No problem. Could you briefly provide a description of what is already done (and maybe how it works) and what needs to be completed?

lupoDharkael commented:
My first comment in this PR contains a very basic overview of the parser but I'll try to describe everything better.

`core/cli_parser.h` contains 2 classes `CommandFlag` and `CommandParser`.

### CommandFlag
Every object of this type contains the information of a command.
I think almost every member variable is self explanatory and I have a few comments too.
The only thing that may create some confusion is the `Vector<CommandChecker *> check_list;`
which is a list of pointer to `CommandChecker`, if you look at the definition of `CommandChecker` you'll see it is just a function that validates arguments. So `check_list` is just a vector of functions to be called by the parser in order to validate the arguments. I thought about changing this to use functions exposed to `ClassDB` so you can define validation functions from gds too.

### CommandParser

Here is a breakdown of the member variables of the class:

- Vector<Ref<CommandFlag> > _commands: the list of CommandFlag available for the parser.

- Vector<String> _game_args: arguments found after a `--`
- Vector<String> _args; arguments found before a `--`
for example, with the arguments `godot clean-the-house 3 -- do-the-dishes`. `_game_args` would contain [do-the-dishes] and `_args` would contain [clean-the-house, 3].

- HashMap<String, String> _data_found: this is where we save the ocurrences of the command flags, the key is the command itself to see if it was wound in the parsed arguments. A key-value example would be <godot clean-the-house, 3> for the example arguments above.

- String _project_file: when we parse one of the first things the parser does is search of a project file as (when I wrote this PR) Godot checks the last argument to be a .godot or scene.

- String _help_header: the text shown before the list of flags in the --help.
- String _help_footer: the text shown after the list of flags in the --help.
- String _version: the text shown by --version.

- bool _separator_enabled: if this is true we split the arguments into `_game_args` and `_args`. When this is false all the arguments go to `_args`. The engine always process `--`, this can be disabled for the convenience of user defined parsers in gds.
- bool _search_project_file: when this is disabled the parser doesn't try to find a project file at the end of the arguments.

### The parse()
This function is concise and I think this doesn't need very much explanation, it does the following:
- split the arguments between `_game_args` and `_args`.
- Try to find the project file if `_search_project_file` is true.
- process the arguments in `_args`, if it finds a non existent one it shows an error. If it finds a defined one checks if it needs an argument and validates the command with all its checkers. Save that command and argument in `_data_found`

### Considerations
- I think now some godot commands can receive multiple arguments, in order to implement that CommandFlag should have a method `int arguments_needed() const` instead of `bool needs_argument() const;` and adapt the parse() method.


- To update this PR you need to check the date of the last commit in this PR and start implementing the changes in Godot's main.cpp from that date.
Here's the list of commits in that file:
https://github.com/godotengine/godot/commits/master/main/main.cpp

- Godot's main.cpp may difer a lot from this in the parts where the commands are processed, before the changes in this PR the command processing was very chaotic as it needed validation and processing of commands in multiple parts of the engine initialization. With this parser the validation is centralized. Keep that in mind and check the actual state of the main.cpp in this PR for examples on how to use the parser class.

Shatur95 commented:
Thank you a lot! I will try. Should I open another PR or or will you add me as a contributor to your fork?

lupoDharkael commented:
I think it would be better to open a new one as this has a lot of comments already and it makes navigation through the feedback a bit harder.
If you have any doubts with anything you can ask me.

Xrayez commented:
I think it's better to take this PR as a reference for implementation, if you attempt to rebase the branch to resolve conflicts then it's easy to miss out potential fixes and enhancements made to `master` so far.

lupoDharkael commented:
It's a lot of work to adapt main.cpp
I think it is better to go through the list of commits that change something on the main and adapt it. I already did that when I resumed the work on this PR after a few months of conflicts and most of the commits are easy to adapt as you only have to touch a few lines.

lupoDharkael commented:
The changes on:
 core/bind/core_bind.cpp
core/bind/core_bind.h
core/os/os.cpp
core/os/os.h 
Are very small, I would start from clean Godot master, add the changes in those 4 files, replace main.cpp with the one in this PR and start adding the changes of individual commits in main starting from the date of my last commit.
That's how I did it the last time.

Shatur95 commented:
Thanks for the tips. I agree with @lupoDharkael and currently updating `main.cpp`.

> I think now some godot commands can receive multiple arguments, in order to implement that CommandFlag should have a method int arguments_needed() const instead of bool needs_argument() const; and adapt the parse() method.

I noticed that `--export` and `--export-debug` already accepts two arguments. How it works in current implementation?

lupoDharkael commented:
This PR only receives a single argument as that's how it was when I wrote it.
As we now require supporting multiple arguments per commands the following elements need to be adapted in cli_parser.h and cli_parser.cpp

- `CommandChecker` needs to be able to check a list of strings instead of a single String.
- `CommandFlag::set_argument_name` defines the name of the argument to show in --help, it should receive a list of names of change it to add_argument_name and append a new one with each call.
- CommandParser::needs_argument and CommandFlag::needs_argument should return an int and renamed accordingly.
- CommandParser::_data_found value should be a list of String instead as a single String as multiple arguments can be owned by a single command.
- CommandParser::get_argument as it returns the argument defined in the flag, it can be changed to return the list of arguments.
- the parse() function obviously.

this change also affects main.cpp but the conversion is more or less straightforward.

Shatur95 commented:
Thank you!

Shatur95 commented:
Still working on it. I refactored code a lot, improved error checking, added usage generation, support for positional arguments (it's better to support it instead of manually checking for scene / script), multiply arguments, default values, required arguments, compound arguments (`-abc`), sticky arguments (`-dvalue`), adjacent arguments (`-d=value`), allowed values lists, any number of flags in single command, support for use with different prefixes (e.g. `/command`, `+command`, `-command`, etc, this can be useful for chat commands parsing, for example) with the ability to disable unneeded syntax (for example disable compound arguments) and other API improvements. 
I currently writing tests for all this stuff. But I didn't make any changes to `main.cpp`. I going to first send a PR with only this class and tests to it (as a class for GDScript, because it can be used as is), and then, after we agree on API and merge it, I will send a separate PR with changes to `main.cpp` to simplify review process. What do you think?

lupoDharkael commented:
That was fast :)
Yeah that's better option right now because the PR will accumulate almost no conflicts that way and it will be simpler to review and merge. The changes seem good, lets see what others think about them.

Shatur95 commented:
I decided to add a couple more features before posting. Finally finished: #44594.
